Audit log CRUD permission failures

johnswanson opened the pull request:
When we run `metabase.api.common/*-check` functions like `read-check` or `create-check`, publish an event when the check fails, along with enough context that, on the other side, the audit log handler can record the relevant event.

johnswanson committed:
Audit log CRUD permission failures

When we run `metabase.api.common/*-check` functions like `read-check` or `create-check`, publish an event when the
check fails, along with enough context that, on the other side, the audit log handler can record the relevant event.

dpsutton commented on test/metabase/api/common_test.clj:
```diff
@@ -85,3 +88,50 @@
+               @*events*)))))
+
+  ;; teardown - underive events so they aren't dispatched in other tests
+  (underive ::permission-failure-event :metabase/event)
```
do these need to be in a finally block?

johnswanson replied:
Good call, done!

johnswanson committed:
allow `audit-log/model-name` to work with RootCollections

`(t2/model root-collection)` returns a Class rather than a keyword - `name` doesn't work here, so we'll use
`getSimpleName` instead.

johnswanson committed:
add a default model-details clause for dashboards

If the event-type is not one that we've explicitly delineated, just return an empty map for the model details.

johnswanson committed:
Migration to fit longer model names in `activity.model`

The `NativeQuerySnippet` or `PermissionsGroupMembership` models, for example, were too long to fit in `VARCHAR(16)`.
I'm doubling the length of this column to fit the longest model I see, with a little extra room.

johnswanson committed:
Rename `RootCollection` to `Collection` in audit logs

johnswanson committed:
fix schema for permission events

Some models we'll report permissions failures on are not actually instances of `:metabase/model`. But as long as they
implement `t2/model`, we can audit log them.

johnswanson commented:
😌 this is ready to review again after the rebase!

johnswanson committed:
do not fire events on read permission failure

These will go to the `read_log` instead of the audit log.

If the event isn't derived from `:metabase/event` it will throw an exception. That `derive` no longer belongs in the
audit log code, and until we have the view log there's not really a logical place to put it. For now, I just commented
out the `publish-event!` call, with the assumption that the view log is coming very soon. Otherwise we can just delete
it and add it back in when it's needed.

noahmoss commented on test/metabase/api/common_test.clj:
```diff
@@ -85,3 +89,49 @@
+  [topic event]
+  (swap! *events* conj [topic event]))
+
+(deftest *-check-functions-publish-events
```
Haven't seen the `*-` prefix before, is that to indicate the test is using a dynamic var instead of the real audit tables?

johnswanson replied:
I was using it as a placeholder, like "[something]-check" but I can see that it's confusing in this context - I'll change the name to just `check-functions-publish-events` instead.

noahmoss replied:
Ohh gotcha

johnswanson commented on resources/migrations/000_migrations.yaml:
```diff
@@ -15989,6 +15989,21 @@ databaseChangeLog:
+            tableName: activity
+            columnName: model
+            newDataType: VARCHAR(32)
+      rollback:
+        - modifyDataType:
+            tableName: activity
+            columnName: model
+            newDataType: VARCHAR(16)
```
Ooh great point, this is actually not good at all - if someone upgrades, puts the longer model names into the audit log, and then downgrades, they'll get an error (something like `ERROR:  value too long for type character varying(16)`). Seems better to just skip the rollback.

noahmoss replied:
Good point

johnswanson committed:
remove rollback for migration

johnswanson committed:
rename check-function-publish-events test
